Fix Events feature for FD1.3 (out-of-order event ids, events feature missing) #6093

kmabeeTT opened the pull request:
Closes #6093 - see 2 commit descriptions, mainly:

1. events feature got dropped in FD1.3 , add it back
2. EnqueueRecordEvent, EnqueueWaitForEvent cmds were not ordered with programs in FD1.3 and those cmds if issued after programs could return to completion queue before the program


Will be beefing up Events tests in follow up PR or in this one if I can finish it before tomorrow, but wanted @DrJessop and @abhullar-tt approval, eyes on this in the meantime.

I think I updated cq_prefetcher correctly for all cases.  Didn't bother optimizing/refactoring code here, just wanted to keep the initial fix "simple"

kmabeeTT committed:
#6093: Set Stall flag on RecordEvent/WaitForEvent cmds for FD1.3+ (whb0)

 - Solves out-of-order event_ids in completion queue assert seen with
   upcoming program+events tests where events issued after program
   would return to completion queue before program.

kmabeeTT committed:
#6093: Add Events feature (RecordEvent/WaitForEvent) handling into FD1.3 prefetcher

 - Events functionality got dropped for whb0 when FD1.3 got merged, this adds it back.

 - Move CQ_COMPLETION_16B_SCRATCH to dedicated addr in common_runtime_address_map.h
   In FD1.3 it was used in prefetcher and it overlapped with CQ_ISSUE_READ_PTR which
   looked ... dangerous. If there is another addr this scratch l1 can overlap with,
   would be great to change and not eat up last 16B slot.

abhullar-tt reviewed (approved):
looks good, just highlighting that we won't be supporting multi-CQ on remote device so  no need to handle `is_event_sync` case for the `REMOTE_PUSH_AND_PULL` path

kmabeeTT committed:
#6093: Add pair of negative Host/Device Event sync tests on future events (expected hangs)

 - Helps verify Events feature is working (spoiler: it wasn't lately)
   by syncing on a future event that is never issued, and ensure a hang
   is seen from Host and from Device (seperate tests)

 - watcher_server_clear_error_flag() is renamed to be
   watcher_server_set_error_flag(bool) for use here.

 - Add watcher_server_killed_due_to_error() support in EventSynchronize
   for host sync negative test to work, and to receive exit signal from
   main thread similar to Finish()
